### PR TITLE
Initial work on Multiparty block commitments

### PR DESCRIPTION
**Pull request by kantai**
### Description

This PR is the first step towards multiparty block commitments, implementing #12 by using a multi-party contract. The multi-party contract checks the transaction sender and the supplied signatures to confirm that at least `required-signers` count of approvals have been achieved before issuing a block-commit.

This PR includes work on the hyperchain miner to support this abstraction: the block commitment implementation will shuttle signatures to the configured multi-party contract. However, the miner isn't yet capable of obtaining those signatures from the other miners. That will require adding a new endpoint for block proposals.

This PR includes two kinds of tests:

1. Unit tests for the multi-party contract via new Clarinet tests
2. An integration test for the miner's implementation of the new block commit

This PR also includes a TOML file which can be used to configure a mocknet L1 node to run as a Stacks 2.1 node, but it requires changes in the `stacks-node` implementation, so that config file isn't used in our integration tests yet.

### Applicable issues
- #12

### CODE REVIEW

**Review by gregorycoppola — APPROVED** [review]
thanks for starting this!

File: core-contracts/tests/hyperchains/multiminer_test.ts [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        };
+
+        const signatory2 = { 
+            secretKey: "003f8c631e98bf52b8dfa36f02df0aaab85dfefc5d8bedb41bc5184afdc4a16001",
+            publicKey: "0294cf0f56b638b2d38c39a92a90692c7ed4eb980832d8dca33cacfeb80c3f2741",
+            address: "STECHMJGSBWNGW3MS334R3PHQD4F59EFMAXY7Y7F"
+        };
+
+        // set the multi_miner_contract as the miner of the hyperchains contract
+        //  and set alice and signatory as miners in the multi-miner contract
```
Comment (pavitthrap) [inline core-contracts/tests/hyperchains/multiminer_test.ts:152]: update comment

File: testnet/stacks-node/src/burnchains/commitment.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        }
+
+        // step 1: figure out the miner's nonce
+        let miner_address = l1_addr_from_signer(self.config.is_mainnet(), op_signer);
+        let nonce = l1_get_nonce(&self.config.get_rpc_url(), &miner_address).map_err(|e| {
+            error!("Failed to obtain miner nonce: {}", e);
+            e
+        })?;
+
+        // step 2: fee estimate (todo: #issue)
```
Comment (pavitthrap) [inline testnet/stacks-node/src/burnchains/commitment.rs:195]: did you want to link an issue here?

File: testnet/stacks-node/src/tests/l1_observer_test.rs [hunk 3]
```diff
@@ -238,6 +238,55 @@ fn select_transactions_where(
...
+    );
+    l1_nonce += 1;
+
+    submit_tx(&l1_rpc_origin, &trait_publish);
+    // Because the nonce ensures that the trait contract is published
+    // before the HC contract, we can broadcast them all at once, even
+    // though the HC contract depends on that contract.
+    submit_tx(&l1_rpc_origin, &hc_contract_publish);
+
+    println!("Submitted FT, NFT, and Hyperchain contracts!");
```
Comment (pavitthrap) [inline testnet/stacks-node/src/tests/l1_observer_test.rs:285]: update print here

File: testnet/stacks-node/src/tests/l1_observer_test.rs [hunk 4]
```diff
@@ -238,6 +238,55 @@ fn select_transactions_where(
     return result;
 }
 
+/// Uses MOCKNET_PRIVATE_KEY_1 to publish the hyperchains contract and supporting
+///  trait contracts
+pub fn publish_hc_contracts_to_l1(mut l1_nonce: u64, config: &Config, miner: PrincipalData) -> u64 {
+    let trait_standard_contract_name = "trait-standards";
+    let l1_rpc_origin = config.burnchain.get_rpc_url();
+    // Publish the NFT/FT traits
```
Comment (pavitthrap) [inline testnet/stacks-node/src/tests/l1_observer_test.rs:246]: update comment

File: core-contracts/contracts/multi-miner.clar [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+(define-private (check-miners (provided-set (list 10 principal)))
+    (let ((provided-checked (filter test-is-none (map index-of-miner provided-set)))
+          (uniques-checked (fold unique-helper provided-checked { all-unique: true, priors: (list)})))
+         (asserts! (get all-unique uniques-checked) (err ERR_SIGNER_APPEARS_TWICE))
+         (asserts! (>= (len provided-checked) signers-required) (err ERR_NOT_ENOUGH_SIGNERS))
+         (ok true)))
+
+;; Remove this function in 2.1
+(define-private (to-consensus-buff (block-data { block: (buff 32), withdrawal-root: (buff 32), target-tip: (buff 32), multi-contract: principal })) (ok 0x00))
```
[the remark this answers is not in the document]
Comment (kantai) [reply core-contracts/contracts/multi-miner.clar:59]: Block height isn't part of the commit data: it's just computed from the current block height at evaluation. The `target-tip` data will prevent block commits from being used on forks or at unintended heights.

File: testnet/stacks-node/src/tests/l1_observer_test.rs (hunk 4 above)
Comment (kantai) [reply testnet/stacks-node/src/tests/l1_observer_test.rs]: Fixed in 8f275f0268b0b5659f2842fa2c7437f2041e30b3

File: testnet/stacks-node/src/tests/l1_observer_test.rs (hunk 3 above)
Comment (kantai) [reply testnet/stacks-node/src/tests/l1_observer_test.rs]: Fixed in 8f275f0268b0b5659f2842fa2c7437f2041e30b3

File: testnet/stacks-node/src/burnchains/commitment.rs (hunk 2 above)
Comment (kantai) [reply testnet/stacks-node/src/burnchains/commitment.rs]: Linked to #140 in 8f275f0268b0b5659f2842fa2c7437f2041e30b3

File: core-contracts/tests/hyperchains/multiminer_test.ts (hunk 1 above)
Comment (kantai) [reply core-contracts/tests/hyperchains/multiminer_test.ts]: Fixed in 8f275f0268b0b5659f2842fa2c7437f2041e30b3

File: core-contracts/contracts/multi-miner.clar [hunk 6]
```diff
@@ -0,0 +1,87 @@
+;; The .hyperchains contract
```
[the remark this answers is not in the document]
Comment (kantai) [reply core-contracts/contracts/multi-miner.clar]: Fixed in 8f275f0268b0b5659f2842fa2c7437f2041e30b3

**Pull request merged**